pkg/storage/gcs.go

```diff
@@ -258,3 +291,76 @@ func hasSSTFiles(ctx context.Context, bucket *storage.BucketHandle, prefix strin
 	}
 	return false
 }
+
+// gcsObjectReader wrap storage.Reader and add the `Seek` method.
+type gcsObjectReader struct {
+	storage   *gcsStorage
+	name      string
+	objHandle *storage.ObjectHandle
+	reader    io.ReadCloser
+	pos       int64
+	// reader context used for implement `io.Seek`
+	// currently, lightning depends on package `xitongsys/parquet-go` to read parquet file and it needs `io.Seeker`
+	// See: https://github.com/xitongsys/parquet-go/blob/207a3cee75900b2b95213627409b7bac0f190bb3/source/source.go#L9-L10
+	ctx context.Context
+}
+
+// Read implement the io.Reader interface.
+func (r *gcsObjectReader) Read(p []byte) (n int, err error) {
+	if r.reader == nil {
+		rc, err := r.objHandle.NewRangeReader(r.ctx, r.pos, -1)
+		if err != nil {
+			return 0, errors.Annotatef(err,
+				"failed to read gcs file, file info: input.bucket='%s', input.key='%s'",
+				r.storage.gcs.Bucket, r.name)
+		}
+		r.reader = rc
+	}
+	n, err = r.reader.Read(p)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	r.pos += int64(n)
+	return n, err
+}
+
+// Close implement the io.Closer interface.
+func (r *gcsObjectReader) Close() error {
+	if r.reader == nil {
+		return nil
+	}
+	return r.reader.Close()
+}
+
+// Seek implement the io.Seeker interface.
+//
+// Currently, tidb-lightning depends on this method to read parquet file for gcs storage.
+func (r *gcsObjectReader) Seek(offset int64, whence int) (int64, error) {
+	var realOffset int64
+	switch whence {
+	case io.SeekStart:
+		realOffset = offset
+	case io.SeekCurrent:
+		realOffset = r.pos + offset
+	case io.SeekEnd:
```

Comment:
I'm not sure if this means reverse offset. (e.g. end - offset)

Reply:
`end + offset` is correct, see https://github.com/golang/go/blob/21a04e33353316635b5f3351e807916f3bb1e844/src/io/example_test.go#L202.

Reply:
Get it.